core/iwasm/common/wasm_shared_memory.c

```diff
@@ -87,6 +88,41 @@ search_module(WASMModuleCommon *module)
     return NULL;
 }
 
+static void
+notify_stale_threads(void *key, void *value, void *user_data)
+{
+    WASMModuleInstanceCommon *module_inst =
+        (WASMModuleInstanceCommon *)user_data;
+    wasm_runtime_atomic_notify(module_inst, key, UINT32_MAX);
+    bh_hash_map_remove(wait_map, key, NULL, NULL);
+}
+
+static void
+create_copy_of_wait_map(void *key, void *value, void *map_copy)
+{
+    bh_hash_map_insert((HashMap *)map_copy, key, value);
+}
+
+void
+notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
+{
+
+    wait_map_without_lock = bh_hash_map_create(
+        32, false, (HashFunc)wait_address_hash,
+        (KeyEqualFunc)wait_address_equal, NULL, destroy_wait_info);
+    /**
+     * create a copy of wait_map with no lock
+     * this allows invoking methods in the callback
+     * which require acquiring lock on the original map
```

Comment:
I don't see there's any lock in the `notify_stale_threads` callback. Should we rather have a lock on the `shared_memory_list_lock` mutex at the beginning of the function and release it at the end? Otherwise the map could be modified in the other thread making the traversal corrupted.

Reply:
the `wasm_runtime_atomic_notify` step in callback calls `acquire_wait_info` which tries to acquire `shared_memory_list_lock`, so I guess encapsulating it with that will cause issues?

Reply:
`HashMap` doesn't duplicate the key/values pairs, so `wait_map_without_lock` will point to the same elements of `wait_map`

Reply:
I think you at least want to lock while you copy hashmaps. Also, what happens if after making a copy a new waiter is added to the map?

Reply:
While copying the original map, the original map already acquires a lock as it traversal. as inserts happen in the copy map, every insert locks the copy map. 
What could be concerning here is that during the traversal of copy map, the original map could be updated. If I put a lock on the original map during traversal, the execution gets stuck because the callback steps in `notify_stale_threads` try to put a lock on the original map. This might lead to some entries (newly added ones) in the original map not catered to at that point of time, but would be eventually handled, right?

Reply:
either way I'm still not sure it would be safe since you're not cloning the map but copying the pointers iiuc